### PR TITLE
[14.0] Shopfloor: Add no_prefill_qty option

### PR DESCRIPTION
**Pull request by mmequignon**
ref: rau-30

### CODE REVIEW

File: shopfloor/services/checkout.py [hunk 1]
```diff
@@ -352,11 +352,14 @@ def select(self, picking_id):
...
 
-    def _select_lines(self, lines):
+    def _select_lines(self, lines, no_prefill_qty=None):
         for line in lines:
             if line.shopfloor_checkout_done:
                 continue
-            line.qty_done = line.product_uom_qty
+            if no_prefill_qty is None:
+                no_prefill_qty = line.product_uom_qty
+            line.qty_done = self._get_prefill_qty(line, no_prefill_qty)
```
Comment (sebalix) [inline shopfloor/services/checkout.py]: I'm not sure to understand this parameter `no_prefill_qty`, is it a boolean or a qty? Just talking about the name, should it be renamed?

File: shopfloor/services/checkout.py [hunk 2]
```diff
@@ -403,12 +411,15 @@ def scan_line(self, picking_id, barcode):
...
+
         if not product:
             packaging = search.packaging_from_scan(barcode)
             if packaging:
-                product = packaging.product_id
-        if product:
-            return self._select_lines_from_product(picking, selection_lines, product)
+                return self._select_lines_from_packaging(
+                    picking, selection_lines, packaging
+                )
```
Comment (sebalix) [inline shopfloor/services/checkout.py]: As you put a `if product` above, we could remove the `if not product` right after to get only:
```python
            packaging = search.packaging_from_scan(barcode)
            if packaging:
                return self._select_lines_from_packaging(
                    picking, selection_lines, packaging
                )
```

File: shopfloor/services/checkout.py (hunk 1 above)
Comment (mmequignon) [reply shopfloor/services/checkout.py]: It is a quantity that represents the quantity to add to `qty_done` when the `no_prefill_qty` option is set to False
Do you think it needs to be renamed? Do you have a suggestion?

File: shopfloor/services/zone_picking.py [hunk 3]
```diff
@@ -765,6 +797,14 @@ def set_destination(
         * set_line_destination+confirmation_required: the scanned location is not
           in the expected one but is valid (in picking type's default destination)
         """
+        if self.work.menu.no_prefill_qty:
+            # While a product, a lot, a packaging …  is scanned, increment the
+            # quantity_done.
+            # Otherwise, fallback to the destination package
+            res = self.scan_source(barcode, quantity, confirmation)
```
Comment (sebalix) [inline shopfloor/services/zone_picking.py]: Why do we call the previous endpoint `scan_source` from `set_destination` here? I do not understand the comment.

File: shopfloor/services/cluster_picking.py [hunk 4]
```diff
@@ -564,6 +617,9 @@ def _scan_line_by_location(self, picking, move_line, location):
         return self._response_for_scan_destination(move_line)
 
     def scan_destination_pack(self, picking_batch_id, move_line_id, barcode, quantity):
+        # TODO rewrite
```
Comment (jbaudoux) [inline shopfloor/services/cluster_picking.py]: Like something is still lacking here?

File: shopfloor/services/location_content_transfer.py [hunk 5]
```diff
@@ -542,6 +545,21 @@ def scan_package(self, location_id, package_level_id, barcode):
...
+        """Either increments qty_done or sets it depending on the `no_prefill_qty`
+        option value.
+        """
+        qty = self._get_prefill_qty(move_line, qty)
+        if self.work.menu.no_prefill_qty:
+            move_line.qty_done += qty
+        else:
+            move_line.qty_done = qty
+
+    def _get_prefill_qty(self, move_line, qty=False):
```
Comment (jbaudoux) [inline shopfloor/services/location_content_transfer.py]: It should not be supported by this scenario as the qty done is always set on all moves at the beginning.
Quantity should be managed when unloading and not when loading, so completely different from what was done on the other scenario.
Should be dropped from this scenario

File: shopfloor/services/zone_picking.py (hunk 3 above)
Comment (jbaudoux) [reply shopfloor/services/zone_picking.py]: The step is calling set_destination on scan but you can continue to scan the source to increment the qty of the scanned source instead of scanning a destination

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: shopfloor/actions/search.py [hunk 6]
```diff
@@ -34,15 +34,11 @@ def picking_from_scan(self, barcode):
             return model.browse()
         return model.search([("name", "=", barcode)], limit=1)
 
-    def product_from_scan(self, barcode, use_packaging=True):
+    def product_from_scan(self, barcode):
```
Comment (simahawk) [inline shopfloor/actions/search.py:37]: I would not remove this on 14.0 to not risk to break things out there.
Please restore this key and log a deprecation msg.

Comment (simahawk) [reply shopfloor/actions/search.py:37]: Although I don't see why we should drop it... can you mention a reason?

File: shopfloor/services/checkout.py [hunk 7]
```diff
@@ -352,22 +352,42 @@ def select(self, picking_id):
...
+    def _select_lines(self, lines, no_prefill_qty=None):
         for line in lines:
             if line.shopfloor_checkout_done:
                 continue
-            line.qty_done = line.product_uom_qty
+            if no_prefill_qty is None:
+                qty = line.product_uom_qty
+            else:
+                qty = no_prefill_qty
+            self._set_or_increment_qty_done(line, qty)
```
Comment (simahawk) [inline shopfloor/services/checkout.py]: `no_prefill_qty` seems to be a flag but is used below as a value.... I would rename it to `prefill_qty` and default it to -1, then
```suggestion
            qty = line.product_uom_qty if prefill_qty > 0 else prefill_qty
            self._set_or_increment_qty_done(line, qty)
```

File: shopfloor/services/checkout.py [hunk 8]
```diff
@@ -352,22 +352,42 @@ def select(self, picking_id):
...
 
     def _deselect_lines(self, lines):
         lines.filtered(lambda l: not l.shopfloor_checkout_done).write(
             {"qty_done": 0, "shopfloor_user_id": False}
         )
 
+    def _set_or_increment_qty_done(self, move_line, qty=False):
+        """Either increments qty_done or sets it depending on the `no_prefill_qty`
+        option value.
+        """
```
Comment (simahawk) [inline shopfloor/services/checkout.py]:
```suggestion
    def _set_qty_done(self, move_line, qty=False):
        """Update qty done on give move line.

        Increase the qty or sets it depending on the `no_prefill_qty` option value.
        """
```

File: shopfloor/services/checkout.py [hunk 9]
```diff
@@ -427,6 +454,8 @@ def _select_lines_from_package(self, picking, selection_lines, package):
                 },
             )
         self._select_lines(lines)
+        if self.work.menu.no_prefill_qty:
+            return self._response_for_select_package(picking, picking.move_line_ids)
```
Comment (simahawk) [inline shopfloor/services/checkout.py]:
```suggestion
        if self.work.menu.no_prefill_qty:
           lines = picking.move_line_ids
```

File: shopfloor/services/checkout.py [hunk 10]
```diff
@@ -463,7 +492,49 @@ def _select_lines_from_product(self, picking, selection_lines, product):
             # package and we have only one.
             return self._select_lines_from_package(picking, selection_lines, packages)
 
-        self._select_lines(lines)
+        self._select_lines(lines, no_prefill_qty=1.0)
+        if self.work.menu.no_prefill_qty:
+            return self._response_for_select_package(picking, picking.move_line_ids)
+        return self._response_for_select_package(picking, lines)
```
Comment (simahawk) [inline shopfloor/services/checkout.py]:
```suggestion
        if self.work.menu.no_prefill_qty:
            lines = picking.move_line_ids
        return self._response_for_select_package(picking, lines)
```

File: shopfloor/services/checkout.py [hunk 11]
```diff
@@ -463,7 +492,49 @@ def _select_lines_from_product(self, picking, selection_lines, product):
...
+                picking, message=self.msg_store.scan_lot_on_product_tracked_by_lot()
+            )
+
+        lines = selection_lines.filtered(lambda l: l.product_id == product)
+        if not lines:
+            return self._response_for_select_line(
+                picking,
+                message={
+                    "message_type": "error",
+                    "body": _("Product is not in the current transfer."),
```
Comment (simahawk) [inline shopfloor/services/checkout.py]: don't we have an exiting message for this?

File: shopfloor/services/checkout.py [hunk 12]
```diff
@@ -463,7 +492,49 @@ def _select_lines_from_product(self, picking, selection_lines, product):
...
+                picking, message=self.msg_store.product_multiple_packages_scan_package()
+            )
+        elif packages:
+            # Select all the lines of the package when we scan a product in a
+            # package and we have only one.
+            return self._select_lines_from_package(picking, selection_lines, packages)
+
+        self._select_lines(lines, no_prefill_qty=packaging.qty)
+        if self.work.menu.no_prefill_qty:
+            return self._response_for_select_package(picking, picking.move_line_ids)
```
Comment (simahawk) [inline shopfloor/services/checkout.py]:
```suggestion
        if self.work.menu.no_prefill_qty:
            lines = picking.move_line_ids
```

File: shopfloor/services/checkout.py [hunk 13]
```diff
@@ -495,7 +566,9 @@ def _select_lines_from_lot(self, picking, selection_lines, lot):
             # package and we have only one.
             return self._select_lines_from_package(picking, selection_lines, packages)
 
-        self._select_lines(lines)
+        self._select_lines(lines, no_prefill_qty=1.0)
+        if self.work.menu.no_prefill_qty:
+            return self._response_for_select_package(picking, picking.move_line_ids)
```
Comment (simahawk) [inline shopfloor/services/checkout.py]:
```suggestion
            lines = picking.move_line_ids
```

File: shopfloor/services/checkout.py [hunk 14]
```diff
@@ -751,6 +824,11 @@ def scan_package_action(self, picking_id, selected_line_ids, barcode):
         * summary: if there is no other lines, go to the summary screen to be able
         to close the stock picking
         """
+        if self.work.menu.no_prefill_qty:
+            res = self.scan_line(picking_id, barcode)
```
Comment (simahawk) [inline shopfloor/services/checkout.py]: comment pls

File: shopfloor/services/cluster_picking.py [hunk 15]
```diff
@@ -456,6 +456,11 @@ def scan_line(self, picking_batch_id, move_line_id, barcode):
         if product and move_line.product_id == product:
             return self._scan_line_by_product(picking, move_line, product)
 
+        if not product:
+            packaging = search.packaging_from_scan(barcode)
+            if packaging and move_line.product_id == product:
```
Comment (simahawk) [inline shopfloor/services/cluster_picking.py]: meaning that `move_line.product_id` is not set? I don't think it can happen... Maybe better w/ `packaging.product_id`. Also, it tells me this case is not tested -> please add a test!

File: shopfloor/services/cluster_picking.py [hunk 16]
```diff
@@ -474,8 +479,27 @@ def scan_line(self, picking_batch_id, move_line_id, barcode):
             move_line, message=self.msg_store.barcode_not_found()
         )
 
+    def _set_or_increment_qty_done(self, move_line, qty=False):
```
Comment (simahawk) [inline shopfloor/services/cluster_picking.py]: rewrite it as suggested above. Also, move these methods to an helper or to a mixin instead of copy/pasting it around everywhere.

File: shopfloor/services/cluster_picking.py (hunk 4 above)
Comment (simahawk) [reply shopfloor/services/cluster_picking.py]: :man_shrugging:

File: shopfloor/services/zone_picking.py [hunk 17]
```diff
@@ -477,7 +477,11 @@ def _scan_source_location(self, barcode, confirmation=False):
             match_user=True,
         )
         if move_lines:
+            move_line = first(move_lines)
             response = self._response_for_set_line_destination(first(move_lines))
+            response["data"]["set_line_destination"]["move_line"][
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py]: this is quite ugly. Please find a better way (eg: pass to all `self._response_for*` a `data_for_move_line` optional arg to extend update the value returned by `self._data_for_move_line`

File: shopfloor/services/zone_picking.py [hunk 18]
```diff
@@ -507,7 +511,11 @@ def _scan_source_package(self, barcode, confirmation=False):
             return response, message
         move_lines = self._find_location_move_lines(package=package)
         if move_lines:
-            response = self._response_for_set_line_destination(first(move_lines))
+            move_line = first(move_lines)
+            response = self._response_for_set_line_destination(move_line)
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py]: ugly as above

File: shopfloor/services/zone_picking.py [hunk 19]
```diff
@@ -535,23 +543,43 @@ def _scan_source_package(self, barcode, confirmation=False):
...
+            # Since there's no recover feature in this scenario, do not write
+            # qty_done on the backend, but send it to the frontend in the response
+            # instead
+            move_line = first(move_lines)
+            response = self._response_for_set_line_destination(move_line)
+            if packaging:
+                qty = packaging.qty
+            else:
+                qty = 1.0
+            response["data"]["set_line_destination"]["move_line"][
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py]: ugly as above

File: shopfloor/services/zone_picking.py [hunk 20]
```diff
@@ -535,23 +543,43 @@ def _scan_source_package(self, barcode, confirmation=False):
...
-            response = self._response_for_set_line_destination(first(move_lines))
+            # Since there's no recover feature in this scenario, do not write
+            # qty_done on the backend, but send it to the frontend in the response
+            # instead
+            move_line = first(move_lines)
+            response = self._response_for_set_line_destination(move_line)
+            if packaging:
+                qty = packaging.qty
+            else:
+                qty = 1.0
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py]:
```suggestion
            qty = packaging.qty or 1.0
```

File: shopfloor/services/zone_picking.py [hunk 21]
```diff
@@ -535,23 +543,43 @@ def _scan_source_package(self, barcode, confirmation=False):
...
+            response["data"]["set_line_destination"]["move_line"][
+                "qty_done"
+            ] = qty_done + self._get_prefill_qty(move_line, qty=qty)
         else:
             response = self._list_move_lines(self.zone_location)
             message = self.msg_store.product_not_found()
         return response, message
 
-    def _scan_source_lot(self, barcode, confirmation=False):
+    def _scan_source_lot(self, barcode, qty_done, confirmation=False):
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py]: if someone has overridden these methods you'll break their code. Make it optional in all methods alike.

File: shopfloor/services/zone_picking.py (hunk 3 above)
Comment (simahawk) [reply shopfloor/services/zone_picking.py]: After seeing this for way too many times I think is ugly as the other hack. 
The comment does not explain what you typed here.

File: shopfloor/tests/test_zone_picking_select_line.py [hunk 22]
```diff
@@ -526,3 +532,102 @@ def test_list_move_lines_empty_location(self):
...
+        self.assert_response_set_line_destination(
+            response,
+            zone_location=self.zone_location,
+            picking_type=self.picking_type,
+            move_line=move_line,
+            qty_done=1.0,
+        )
+
+    def test_scan_source_barcode_packaging_no_prefill(self):
+        """When a packaging is scanned, qty_done in response is
```
Comment (simahawk) [inline shopfloor/tests/test_zone_picking_select_line.py]: shorten this

File: shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js [hunk 23]
```diff
@@ -112,10 +113,35 @@ export var PackagingQtyPickerMixin = {
...
+        },
     },
     created: function () {
+        this.qty_done = parseInt(this.opts.qty_done, 10);
         this.qty_todo = parseInt(this.opts.init_value, 10);
-        this.qty_done = parseInt(this.opts.init_value, 10);
+        this.qty_todo_by_pkg = this.compute_qty_todo_by_pkg();
+    },
+    updated: function () {
+        if (this.$route.params.state !== "change_quantity") {
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js]: this is specific to a scenario :red_circle:

File: shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js [hunk 24]
```diff
@@ -266,6 +294,9 @@ export var PackagingQtyPicker = Vue.component("packaging-qty-picker", {
                             @blur="$event.target.value=qty_by_pkg[pkg.id]"
                             />
                     </v-col>
+                    <v-col cols="1" md="2">
+                        <span class="qty-todo">/ {{ get_qty_todo_by_pkg(pkg.id) }}</span>
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js]: I'd like to see a screenshot of the result

File: shopfloor_mobile/static/wms/src/scenario/checkout.js [hunk 25]
```diff
@@ -65,6 +65,7 @@ const Checkout = {
                     :select_records="state.data.picking.move_lines"
                     :select_records_grouped="utils.wms.group_lines_by_location(state.data.picking.move_lines, {'prepare_records': utils.wms.only_one_package})"
                     :select_options="select_line_manual_select_opts()"
+                    :use_qty_todo_by_pkg="true"
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/scenario/checkout.js]: as per select options I would encapsulate this setting for the qty picker into `:qty_picker_options`

File: shopfloor_mobile/static/wms/src/scenario/cluster_picking.js [hunk 26]
```diff
@@ -34,8 +34,11 @@ const ClusterPicking = {
                 v-if="state_in(['start_line', 'scan_destination', 'change_pack_lot', 'stock_issue'])"
                 :line="state.data"
                 :article-scanned="state_is('scan_destination')"
-                :show-qty-picker="state_is('scan_destination')"
+                :show_qty_picker="state_is('scan_destination')"
+                :show_qty_picker_display="state_is('start_line')"
+                :use_qty_todo_by_pkg="true"
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/scenario/cluster_picking.js]: `:qty_picker_options`

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by jbaudoux** [comment]
@JuMiSanAr

## Zone picking

### 1/ Scan location - wrong packaging qty initialized
- [x] The quantity for the packaging is not initialized when you start by scanning a location:
image
The packaging is btw also wronly displayed. It should be 3 transport boxes & 0 units

### 2/ Location, then product Internal error
- [x] Scan the location to open the line
Scan the product to increase the quantity
```
  File "wms/shopfloor/services/zone_picking.py", line 813, in set_destination
    accept_only_package = not self._move_line_full_qty(move_line, quantity)
  File "wms/shopfloor/services/zone_picking.py", line 707, in _move_line_full_qty
    move_line.product_uom_qty - qty, precision_rounding=rounding
TypeError: unsupported operand type(s) for -: 'float' and 'NoneType'
```

### 3/ Wrong package lost quantity + internal error
- [x] Have a line with qty > 1
Scan the right product to open the line.
Scan a wrong product (a barcode of another existing product)
You get a message but the previous quantity is lost
Scan again the right product, you get:
```
  File "wms/shopfloor/services/zone_picking.py", line 636, in scan_source
    response, message = handler(barcode, qty_done, confirmation=confirmation)
  File "wms/shopfloor/services/zone_picking.py", line 575, in _scan_source_product
    ] = qty_done + self._get_prefill_qty(move_line, qty=qty)
TypeError: unsupported operand type(s) for +: 'NoneType' and 'float'
```

### 4/ Edit quantity issue
- [ ] When you are processing a line, I cannot change anymore the content of any quantity fields

### 5/ Location scan should not increase quantity
- [ ] When you scan the location to open the product to process, the quantity should be initialized to 0

### 6/ Scan product update the quantity
- [ ] When you scan the product, the quantity is increased in the backend but not on the front-end


## Cluster picking
### 1/ Readonly issue

- [x] When the line to process is shown with quantity widget in readonly, the packaging info has been lost.
Before:
image
With this PR, no prefill option inactive
image
With this PR, no prefill option active
image

### 2/ Edit quantity issue
- [x] On the next screen once you have selected the product, I cannot change anymore the content of any quantity fields

### 3/ Display issue
- [ ] Can you improve the display to prevent line return in the packaging quantity
image

## Checkout
### 1/ Increased scanned quantity
- [ ] When the product is in a source package (e.g. 2 units in a PACK01), each time I scan the product the quantity is increased by the total quantity (i.e. 2) instead of 1

### 2/ Coloring
- [x] Lines that are not processed are white, lines that are properly processed are green.
Now that you increase step by step the quantity, the line should be colored differently is taken < requested or taken > requested, like on the quantity widget where you have a light red or strong red on the quantity input

### 3/ Edit quantity and split
- [ ] If I edit the quantity with the pencil and decrease the quantity, the line is split in 2 while only the taken quantity (done quantity) should be decreased. So if the source & destination pack is the same, no split should occur.
Before:
ProdA | src pkg: PACK01 | dest pck: PACK01 | reserved: 2 | done: 2
After quantity edit, expected:
ProdA | src pkg: PACK01 | dest pck: PACK01 | reserved: 2 | done: 1
and not:
ProdA | src pkg: PACK01 | dest pck: | reserved: 1 | done: 0
ProdA | src pkg: PACK01 | dest pck: PACK01 | reserved: 1 | done: 1

File: shopfloor/services/checkout.py (hunk 1 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: It has been changed into prefill_qty

File: shopfloor/services/checkout.py (hunk 2 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: done

File: shopfloor/services/checkout.py (hunk 7 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: done

File: shopfloor/services/checkout.py (hunk 8 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: done

File: shopfloor/services/checkout.py (hunk 9 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: done

File: shopfloor/services/checkout.py (hunk 10 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: done

File: shopfloor/services/checkout.py (hunk 11 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: Not exactly, so I created a new one

File: shopfloor/services/checkout.py (hunk 12 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: done

File: shopfloor/services/checkout.py (hunk 13 above)
Comment (TDu) [reply shopfloor/services/checkout.py]: done

**Comment by TDu** [comment]
Rebased on the 14.0 branch

File: shopfloor/services/zone_picking.py (hunk 17 above)
Comment (TDu) [reply shopfloor/services/zone_picking.py]: done

File: shopfloor/services/zone_picking.py (hunk 18 above)
Comment (TDu) [reply shopfloor/services/zone_picking.py]: done

File: shopfloor/services/zone_picking.py (hunk 3 above)
Comment (TDu) [reply shopfloor/services/zone_picking.py]: It should be the job of the front end to call the correct endpoint, IMO.

File: shopfloor/actions/search.py (hunk 6 above)
Comment (TDu) [reply shopfloor/actions/search.py:37]: It is dropped because now it is handled by the `packaging_from_scan` in the specific cases.

File: shopfloor/tests/test_cluster_picking_scan_line.py [hunk 27]
```diff
@@ -404,3 +404,49 @@ def test_scan_line_error_not_found(self):
...
+        )
+        qty_done = response["data"]["scan_destination"]["qty_done"]
+        self.assertEqual(qty_done, expected_qty_done)
+
+    def test_scan_line_pack_no_prefill_ok(self):
+        """Check qty done when pack is scanned and no_prefill_qty is enabled"""
+        self._enable_no_prefill()
+        self._simulate_batch_selected(self.batch, in_package=True)
+        line = self.batch.picking_ids.move_line_ids
+        self._assert_qty_done(line, line.package_id.name, 0.0)
```
Comment (TDu) [inline shopfloor/tests/test_cluster_picking_scan_line.py]: Shouldn't it be ?
```suggestion
        self._assert_qty_done(line, line.package_id.name, 1.0)
```

File: shopfloor/services/cluster_picking.py [hunk 28]
```diff
@@ -456,6 +456,11 @@ def scan_line(self, picking_batch_id, move_line_id, barcode):
         if product and move_line.product_id == product:
             return self._scan_line_by_product(picking, move_line, product)
 
+        if not product:
+            packaging = search.packaging_from_scan(barcode)
```
Comment (TDu) [inline shopfloor/services/cluster_picking.py:460]: Can we have a packaging without a package (search for on line 450) because if the answer is no this will line will never be reached !

Comment (jbaudoux) [inline shopfloor/services/cluster_picking.py]: yes, that's the product packaging, like the barcode on the box (from the manufacturer and set and product master data)

Comment (TDu) [reply shopfloor/services/cluster_picking.py]: oops, sorry I deleted my comment before seeing your answer. Which I thank you for.
it was (can we have a packaging without a package searched 10 lines above)

File: shopfloor/services/cluster_picking.py (hunk 15 above)
Comment (TDu) [reply shopfloor/services/cluster_picking.py]: Fixed

File: shopfloor/services/cluster_picking.py (hunk 28 above)
Comment (TDu) [reply shopfloor/services/cluster_picking.py]: fixed

File: shopfloor/services/location_content_transfer.py (hunk 5 above)
Comment (TDu) [reply shopfloor/services/location_content_transfer.py]: It has been removed

File: shopfloor/services/zone_picking.py (hunk 19 above)
Comment (TDu) [reply shopfloor/services/zone_picking.py]: fixed

File: shopfloor/services/zone_picking.py (hunk 20 above)
Comment (TDu) [reply shopfloor/services/zone_picking.py]: done

File: shopfloor/services/cluster_picking.py (hunk 4 above)
Comment (TDu) [reply shopfloor/services/cluster_picking.py]: I think this is a left over comment because all the quantity increments are made in the _scan_line_by.... and it should be removed.

File: shopfloor/services/cluster_picking.py [hunk 29]
```diff
@@ -583,6 +637,13 @@ def scan_destination_pack(self, picking_batch_id, move_line_id, barcode, quantit
         have the same destination.
         * start_line: to pick the next line if any.
         """
+        if self.work.menu.no_prefill_qty:
+            res = self.scan_line(picking_batch_id, move_line_id, barcode)
+            # If we got an error from scan line, fall back to finding
+            # the destination package
+            message_type = res.get("message", {}).get("message_type")
+            if message_type != "error":
+                return res
```
Comment (TDu) [inline shopfloor/services/cluster_picking.py:646]: I see no relation to this change and the specs ?
Commenting this code does not make the tests fail ?

Comment (TDu) [inline shopfloor/services/cluster_picking.py]: I see no relation to this change and the specs ?
Commenting this code does not make the tests fail ?

Comment (TDu) [reply shopfloor/services/cluster_picking.py]: That is because the test are all based on `scan_line`

File: shopfloor/services/cluster_picking.py [hunk 30]
```diff
@@ -501,6 +523,34 @@ def _scan_line_by_product(self, picking, move_line, product):
...
+        # Do not use mapped here: we want to see if we have more than one package,
+        # but also if we have one product as a package and the same product as
+        # a unit in another line. In both cases, we want the user to scan the
+        # package.
+        if packages and len({line.package_id for line in other_product_lines}) > 1:
+            return self._response_for_start_line(
+                move_line,
+                message=self.msg_store.product_multiple_packages_scan_package(),
+            )
+        self._set_or_increment_qty_done(move_line, packaging.qty)
```
Comment (TDu) [inline shopfloor/services/cluster_picking.py]: This function will modify the qty_done on the move_line ! This is happening before any work is done by the user. This will lead to errors

**Comment by jbaudoux** [comment]
I also included here the fix for the sorting of the packaging when you have a packaging of qty=1.
And I also fixed the qty reserved & available in the picking info display.

**Comment by jbaudoux** [comment]
@TDu This test is failing test_scan_package_action_scan_product2_to_increment_qty
I have disabled it but needs fixing when you tackle checkout scenario issues

**Comment by simahawk** [comment]
I don't have time to get back for a review. I trust JE in this case :wink: 
In any case, please, do not forget to squash and cleanup commits (eg: "removed pdb" :stuck_out_tongue_winking_eye: ).

File: shopfloor/actions/search.py (hunk 6 above)
Comment (simahawk) [inline shopfloor/actions/search.py:37]: why this change?

File: shopfloor/data/shopfloor_scenario_data.xml [hunk 31]
```diff
@@ -1,4 +1,4 @@
-<odoo noupdate="1">
+<odoo noupdate="0">
```
Comment (simahawk) [inline shopfloor/data/shopfloor_scenario_data.xml:1]: Do not change this pls.
Use a migration step if needed.

File: shopfloor/services/checkout.py [hunk 32]
```diff
@@ -352,11 +352,16 @@ def select(self, picking_id):
...
 
-    def _select_lines(self, lines):
+    def _select_lines(self, lines, prefill_qty=0):
         for line in lines:
             if line.shopfloor_checkout_done:
                 continue
-            line.qty_done = line.product_uom_qty
+            if self.work.menu.no_prefill_qty:
+                line.qty_done += prefill_qty
+                # For prefill quantity we only want to increment one line
```
Comment (simahawk) [inline shopfloor/services/checkout.py:361]: I would not change the initial value in any case. I'd use `for i, line in enumerate(lines)` and do it only for i==0 when needed. 

Yet, I don't get why it has to be done on one line only but I might miss some implementation details.

File: shopfloor/services/checkout.py [hunk 33]
```diff
@@ -662,6 +706,25 @@ def _switch_line_qty_done(self, picking, selected_lines, switch_lines):
                 lambda l: l.product_uom_qty,
             )
 
+    def _increment_custom_qty(
+        self, picking, selected_line, increment_lines, qty_increment
+    ):
+        """Increment the  qty_done of a move line with a custom value
+
+        The selected_line_ids parameter is used to keep the selection of lines
```
Comment (simahawk) [inline shopfloor/services/checkout.py:714]: `selected_line_ids` is nowhere to be seen

File: shopfloor/services/cluster_picking.py [hunk 34]
```diff
@@ -100,8 +100,11 @@ def _response_for_start_line(self, move_line, message=None, popup=None):
             popup=popup,
         )
 
-    def _response_for_scan_destination(self, move_line, message=None):
-        data = self._data_move_line(move_line)
+    def _response_for_scan_destination(self, move_line, message=None, qty_done=None):
+        if qty_done is None:
+            data = self._data_move_line(move_line)
+        else:
+            data = self._data_move_line(move_line, qty_done=qty_done)
```
Comment (simahawk) [inline shopfloor/services/cluster_picking.py:107]:
```suggestion
        data = self._data_move_line(move_line, qty_done=qty_done)
```

File: shopfloor/services/cluster_picking.py [hunk 35]
```diff
@@ -474,6 +480,14 @@ def scan_line(self, picking_batch_id, move_line_id, barcode):
             move_line, message=self.msg_store.barcode_not_found()
         )
 
+    def _get_prefill_qty(self, move_line, qty=False):
+        """Returns the increment qty or qty_done depending on the no_prefill_qty
```
Comment (simahawk) [inline shopfloor/services/cluster_picking.py:484]: pls follow this convention for multiline docstrings https://peps.python.org/pep-0257/#multi-line-docstrings

File: shopfloor/services/location_content_transfer.py [hunk 36]
```diff
@@ -504,6 +504,9 @@ def scan_package(self, location_id, package_level_id, barcode):
         other_move_lines = move_lines - package_move_lines
 
         product = search.product_from_scan(barcode)
+        if not product:
```
Comment (simahawk) [inline shopfloor/services/location_content_transfer.py:507]: these changes should be done in different commits...

File: shopfloor/services/zone_picking.py [hunk 37]
```diff
@@ -211,10 +211,12 @@ def _response_for_set_line_destination(
         move_line,
         message=None,
         confirmation_required=False,
+        **kw,
     ):
         if confirmation_required and not message:
             message = self.msg_store.need_confirmation()
         data = self._data_for_move_line(move_line)
+        data["move_line"].update(kw)
```
Comment (simahawk) [inline shopfloor/services/zone_picking.py:219]: why not passing `kw` to `_data_for_move_line`?

File: shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py [hunk 38]
```diff
@@ -0,0 +1,107 @@
...
+            move_line,
+            {move_line: origin_qty_done + 1},
+        )
+
+    def test_scan_package_action_scan_product2_to_increment_qty(self):
+        """
+        two lines with the same product
+
+        """
+        # FIXME - test to be fixed
```
Comment (simahawk) [inline shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py:42]: ?

File: shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js [hunk 39]
```diff
@@ -1,28 +1,39 @@
 /**
  * Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
  * @author Simone Orsi <simahawk@gmail.com>
- * Copyright 2021 Jacques-Etienne Baudoux (BCIM)
+ * Copyright 2021-2022 Jacques-Etienne Baudoux (BCIM)
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js:4]: please move these changes to a specific commit, better, to a specific PR

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by TDu** [comment]
To restore (kind of) atomic commits I reopened a new version of this **https://github.com/OCA/wms/pull/483**
Where I will also attend the last comments from @simahawk

File: shopfloor/actions/search.py (hunk 6 above)
Comment (TDu) [reply shopfloor/actions/search.py:37]: It is dropped because now it is handled by the packaging_from_scan in the specific cases.

File: shopfloor/services/checkout.py (hunk 32 above)
Comment (TDu) [reply shopfloor/services/checkout.py:361]: Good point about not changing the initial value.

File: shopfloor/services/checkout.py (hunk 33 above)
Comment (TDu) [reply shopfloor/services/checkout.py:714]: typo fixed

File: shopfloor/services/cluster_picking.py (hunk 34 above)
Comment (TDu) [reply shopfloor/services/cluster_picking.py:107]: If the parameter is present in the call the `qty_done` in the data will be changed. We don't want that

File: shopfloor/services/cluster_picking.py (hunk 35 above)
Comment (TDu) [reply shopfloor/services/cluster_picking.py:484]: done

File: shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py (hunk 38 above)
Comment (TDu) [reply shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py:42]: activated

Comment (TDu) [reply shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py:42]: If I am not mistaken this test is not relevant anymore here (maybe implemented next) @jbaudoux

File: shopfloor/actions/search.py (hunk 6 above)
Comment (sebalix) [reply shopfloor/actions/search.py:37]: The story is:

1. initially `product_from_scan` was searching both on product and packaging (without any choice)
2. when we introduced the scan of packaging as a dedicated flow in the `delivery` scenario, I added `packaging_from_scan` and needed to add the `use_packaging` option on `product_from_scan` to disable it without breaking other scenarios.

Now if we are sure we are not breaking anything by removing this parameter... OK to remove it and use `packaging_from_scan` in other scenarios to make things clear.

Comment (TDu) [reply shopfloor/actions/search.py:37]: All instance of it have been changed.

**Comment by sebalix** [comment]
I close this one then.

File: shopfloor/actions/search.py (hunk 6 above)
Comment (sebalix) [reply shopfloor/actions/search.py:37]: If the goal (not of this PR, but the general idea) is that all scenarios use this scan handler, we do not solve anything by scanning the product+packaging in the same place where some scenarios have a different flow regarding the packaging (like checkout).
We can keep it like this but we need to be aware of that.
